test/subscription_map.hpp

```diff
@@ -439,31 +439,39 @@ class multiple_subscription_map
     // Insert a key => value at the specified subscription path
     // returns the handle and true if key was inserted, false if key was updated
     template <typename V>
-    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key const &key, V&& value) {
+    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key&& key, V&& value) {
         auto path = this->find_subscription(subscription);
         if (path.empty()) {
             auto new_subscription_path = this->create_subscription(subscription);
-            new_subscription_path.back()->second.value[key] = std::forward<V>(value);
+            new_subscription_path.back()->second.value.insert(std::make_pair(std::forward<Key>(key), std::forward<V>(value)));
             ++this->map_size;
             return std::make_pair(this->path_to_handle(new_subscription_path), true);
         }
         else {
-            auto new_pair = std::make_pair(key, std::forward<V>(value));
-            auto insert_result = path.back()->second.value.insert(new_pair);
-            if(insert_result.second) {
+            // I have to do a find first, because insert_or_assign is available in C++17 only
+            // https://www.fluentcpp.com/2018/12/11/overview-of-std-map-insertion-emplacement-methods-in-cpp17/
+            auto& subscription_set = path.back()->second.value;
+            auto existing_iter = subscription_set.find(key);
+            if(existing_iter == subscription_set.end()) {
+                subscription_set.insert(std::make_pair(std::forward<Key>(key), std::forward<V>(value)));
```

Comment:
emplace?

In fact, I recommend using emplace, *instead* of find, because either the emplacement succeeded, and you're done, or the emplacement failed, and now you have the iterator to the existing item and you can modify it's value directly.